File: typescript/src/code-smells/oop-abusers/temporal-instance-variables.ts
export class PizzaOrder {
  private size?: 'S' | 'M' | 'L' | undefined
  private toppings: string[] = []
  private address?: string | undefined

  start(size: 'S' | 'M' | 'L') {
    this.size = size
    this.toppings = []
    this.address = undefined
  }

  addTopping(topping: string) {
    if (!this.size) {
      return
    }
    this.toppings.push(topping)
  }

  setDeliveryAddress(address: string) {
    this.address = address
  }

  place(): string {
    const summary = `Pizza ${this.size ?? '?'} to ${this.address ?? 'UNKNOWN'} with [${this.toppings.join(', ')}]`
    this.size = undefined
    this.address = undefined
    this.toppings = []
    return summary
  }
}

export function demoPizzaOrder(): string {
  const o = new PizzaOrder()
  o.start('L')
  o.addTopping('pepperoni')
  o.addTopping('mushroom')
  o.setDeliveryAddress('123 Main St')
  return o.place()
}

File: csharp/src/code-smells/oop-abusers/TemporalInstanceVariables.cs
namespace RefactoringAvanzado.CodeSmells.OopAbusers;

public class PizzaOrder
{
    private string? _size;
    private List<string> _toppings = [];
    private string? _address;

    public void Start(string size)
    {
        _size = size;
        _toppings = [];
        _address = null;
    }

    public void AddTopping(string topping)
    {
        if (_size == null)
        {
            return;
        }
        _toppings.Add(topping);
    }

    public void SetDeliveryAddress(string address)
    {
        _address = address;
    }

    public string Place()
    {
        var summary = $"Pizza {_size ?? "?"} to {_address ?? "UNKNOWN"} with [{string.Join(", ", _toppings)}]";
        _size = null;
        _address = null;
        _toppings = [];
        return summary;
    }
}

public static class TemporalInstanceVariables
{
    public static string DemoPizzaOrder()
    {
        var o = new PizzaOrder();
        o.Start("L");
        o.AddTopping("pepperoni");
        o.AddTopping("mushroom");
        o.SetDeliveryAddress("123 Main St");
        return o.Place();
    }
}
